java/serving/src/main/java/feast/serving/service/OnlineServingServiceV2.java

```diff
@@ -255,6 +246,84 @@ private List<Map<String, ValueProto.Value>> getEntityRows(
     return entityRows;
   }
 
+  private List<List<feast.storage.api.retriever.Feature>> retrieveFeatures(
+      List<FeatureReferenceV2> featureReferences, List<Map<String, ValueProto.Value>> entityRows) {
+    // Prepare feature reference to index mapping. This mapping will be used to arrange the
+    // retrieved features to the same order as in the input.
+    if (featureReferences.isEmpty()) {
+      throw new RuntimeException("Requested features list must not be empty.");
+    }
+    Map<FeatureReferenceV2, Integer> featureReferenceToIndexMap =
+        new HashMap<>(featureReferences.size());
+    for (int i = 0; i < featureReferences.size(); i++) {
+      FeatureReferenceV2 featureReference = featureReferences.get(i);
+      if (featureReferenceToIndexMap.containsKey(featureReference)) {
+        throw new RuntimeException(
+            String.format(
+                "Found duplicate features %s:%s.",
+                featureReference.getFeatureViewName(), featureReference.getFeatureName()));
+      }
+      featureReferenceToIndexMap.put(featureReference, i);
+    }
+
+    // Create placeholders for retrieved features.
+    List<List<feast.storage.api.retriever.Feature>> features = new ArrayList<>(entityRows.size());
+    for (int i = 0; i < entityRows.size(); i++) {
+      List<feast.storage.api.retriever.Feature> featuresPerEntity =
+          new ArrayList<>(featureReferences.size());
+      for (int j = 0; j < featureReferences.size(); j++) {
+        featuresPerEntity.add(null);
+      }
+      features.add(featuresPerEntity);
+    }
+
+    // Group feature references by join keys.
+    Map<String, List<FeatureReferenceV2>> groupNameToFeatureReferencesMap =
+        featureReferences.stream()
```

Comment:
To speed up this part we may want to extract distinct feature views from all feature references. And then group feature views instead.

Reply:
IIUC grouping by join keys results in the same or less groups (therefore same or more efficient) than grouping by feature view. The is because multiple feature views can have the same join keys. In [L286](https://github.com/feast-dev/feast/pull/2835/files#diff-ff604cedff2831dc1db08b017aab8dce07a91e20be23dd26787e468da8a1e845R286), `this.registryRepository.getEntitiesList(featureReference)` internally gets feature view spec first, then gets entity names of the feature view spec, then we find join keys for the entity names.

Actually, I grouped by feature view in the beginning. Then I changed to group by join keys in the [second commit](https://github.com/feast-dev/feast/pull/2835/commits/2c9f702756a4313218d78ea51007e753dae12170) of this PR, as an optimization.